File: NeetCodeIO/543DiameterOfBinaryTree/DiameterOfBinaryTree.cpp
class Solution {
public:
    int ans = 0;
    int dfs(TreeNode* root) {
        if (!root) {
            return -1;
        }
        int left = dfs(root->left);
        int right = dfs(root->right);
        ans = max(ans, left + right + 2);
        return 1 + max(left, right);
    }
    int diameterOfBinaryTree(TreeNode* root) {
        // DFS -> Time:O(n), Space:O(n)
        dfs(root);
        return ans;
    }
};

File: NeetCodeIO/543DiameterOfBinaryTree/DiameterOfBinaryTree.py
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        # DFS -> Time:O(n), Space:O(n)
        ans = [0]
        def dfs(root):
            if not root:
                return -1
            left = dfs(root.left)
            right = dfs(root.right)
            ans[0] = max(ans[0], 2 + left + right)
            return 1 + max(left, right)
        dfs(root)
        return ans[0]
